ddt.py

```diff
@@ -18,59 +20,398 @@
 
 
+def ddt(cls):
+    """
+    Class decorator for subclasses of ``unittest.TestCase``.
+
+    Apply this decorator to the test case class, and then decorate test methods
+    with ``@data``, ``@file_data``, ``@unpack``, and ``@unpackall``.
+
+    For each method decorated with ``@data`` and ``@file_data``, this will
+    effectively create one method for each member of the Cartesian product of
+    data items passed to ``@data`` or read from a JSON file by ``@file_data``.
+
+    The names of the test methods follow the pattern
+    ``original_test_name(__{ordinal}_{data})+``.
+
+    The part ``__{ordinal}_{data}`` is repeated as many times as there are
+    nested ``@data`` and ``@file_data`` decorators.
+
+    ``ordinal`` is the position of a particular data item in the list of
+    options represented by the corresponding decorator. Positions are numbered
+    from 0. Keyword arguments (for ``@data``) and members of top-level dict
+    (for ``@data_file``) are ordered according to their keys.
+
+    ``data`` attempts to provide human readable identification of the data item
+    with the following priority: 1) The keyword is used for data items passed
+    as keyword arguments (for ``@data``) or members of the top-level dict (for
+    ``@file_data``). 2) The ``__name__`` attribute of the value if it exists.
+    3) An explicit namestring representation of the data value converted into a
+    valid python identifier, if possible. 4) Only the position is used.
+
+    For each method decorated with ``@file_data('test_data.json')``, the
+    decorator will try to load the test_data.json file located relative
+    to the python file containing the method that is decorated. It will,
+    for each ``test_name`` key create as many methods in the list of values
+    from the ``data`` key.
+
+    """
+    for name, func in list(cls.__dict__.items()):
+        if hasattr(func, PARAMS_SETS_ATTR):
+            test_vectors = itertools.product(*(
+                map(
+                    lambda s: s.use_class(cls),
+                    getattr(func, PARAMS_SETS_ATTR)
+                )
+            ))
+            for vector in test_vectors:
+                params = sum(vector, Params(name, [], {}))
+                params.unpack(getattr(func, UNPACKALL_ATTR, 0))
+                add_test(cls, func, params)
+            delattr(cls, name)
+
+    return cls
+
+
+# Adding new tests
+
+
+def add_test(cls, func, params):
+    """Add a test derived from an original function and specific combination of
+    values provided by ``@data`` and ``@file_data`` decorators.
+
+    """
+    if isinstance(params, Params):
+        @wraps(func)
+        def test_case_func(self):
+            return func(self, *params.args, **params.kwargs)
+
+    else:
+        def test_case_func(self):
+            raise params.reason
+
+    test_case_func.__name__ = params.name
+    setattr(cls, params.name, test_case_func)
+
+
+# Internal data structures
+
+
+class ParamsFailure:
```

Comment:
Correct. I guess it is obvious I am not too fluent n Python 2. :-) 
6ec2c876bf5c400e03f951601d0a87eae15b7b37